File: executor/src/main/java/de/pfadfinden/report_engine/executor/Observability/Logger.java
package de.pfadfinden.report_engine.executor.Observability;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.logs.LogRecordBuilder;
import io.opentelemetry.api.logs.Severity;

/**
 * Structured events - both audit events (report trigger/execution/download, at INFO/ERROR) and
 * operational/debugging ones (cache hits, retries, connection issues, at DEBUG/WARN) - emitted
 * directly through the OTel Logs Bridge API, the same signal pipeline as traces/metrics, exported
 * wherever azure-report-executor/local-report-executor's Telemetry bootstrap sends logs. Not cached
 * in a static field: GlobalOpenTelemetry's logs bridge is fetched fresh on each call so this works
 * correctly regardless of whether the SDK has been registered yet.
 *
 * <p>LOG_LEVEL (debug|info|warn|error, default info) sets the minimum severity actually emitted -
 * DEBUG-level operational detail stays quiet by default so it doesn't crowd out the audit trail,
 * and is available on demand for troubleshooting without a code change.
 */
public final class Logger {

  private static final String SCOPE = "report-engine";
  private static final AttributeKey<String> EVENT_NAME = AttributeKey.stringKey("event.name");
  private static final AttributeKey<String> ERROR_TYPE = AttributeKey.stringKey("error.type");
  private static final AttributeKey<String> ERROR_MESSAGE = AttributeKey.stringKey("error.message");

  private Logger() {}

  /** Operational/debugging detail (e.g. a cache hit, a retry) - quiet unless LOG_LEVEL=debug. */
  public static void debug(String eventName, Attributes attributes) {
    emit(Severity.DEBUG, eventName, attributes, null);
  }

  /** A business-level audit event (e.g. a report was triggered/downloaded). */
  public static void event(String eventName, Attributes attributes) {
    emit(Severity.INFO, eventName, attributes, null);
  }

  /** Something recoverable went wrong or looks off (e.g. a fallback kicked in). */
  public static void warn(String eventName, Attributes attributes) {
    emit(Severity.WARN, eventName, attributes, null);
  }

  /** A business-level audit event that failed. */
  public static void error(String eventName, Attributes attributes, Throwable error) {
    emit(Severity.ERROR, eventName, attributes, error);
  }

  private static void emit(
      Severity severity, String eventName, Attributes attributes, Throwable error) {
    if (severity.getSeverityNumber() < minSeverity().getSeverityNumber()) {
      return;
    }
    LogRecordBuilder builder =
        otelLogger()
            .logRecordBuilder()
            .setSeverity(severity)
            .setBody(eventName)
            .setAttribute(EVENT_NAME, eventName)
            .setAllAttributes(attributes);
    if (error != null) {
      builder
          .setAttribute(ERROR_TYPE, error.getClass().getName())
          .setAttribute(ERROR_MESSAGE, String.valueOf(error.getMessage()));
    }
    builder.emit();
  }

  private static Severity minSeverity() {
    String level = System.getenv("LOG_LEVEL");
    if (level == null) {
      return Severity.INFO;
    }
    return switch (level.toUpperCase()) {
      case "DEBUG" -> Severity.DEBUG;
      case "WARN" -> Severity.WARN;
      case "ERROR" -> Severity.ERROR;
      default -> Severity.INFO;
    };
  }

  private static io.opentelemetry.api.logs.Logger otelLogger() {
    return GlobalOpenTelemetry.get().getLogsBridge().loggerBuilder(SCOPE).build();
  }
}

File: frontend/src/telemetry/logger.ts
import { logs, SeverityNumber } from '@opentelemetry/api-logs';

const SCOPE = 'report-engine';
const EVENT_NAME_ATTRIBUTE = 'event.name';

const LEVELS: Record<string, SeverityNumber> = {
  debug: SeverityNumber.DEBUG,
  info: SeverityNumber.INFO,
  warn: SeverityNumber.WARN,
  error: SeverityNumber.ERROR,
};

type AuditAttributes = Record<string, string | number | boolean>;

/**
 * Structured events - both audit events (report trigger, at INFO/ERROR) and operational/debugging
 * ones (cache hits, download retries, at DEBUG/WARN) - emitted directly through the OTel Logs
 * API, the same signal pipeline telemetry.ts wires up. The Java-side counterpart is
 * executor's Observability/Logger.java - see that class for the full rationale (same emission
 * shape, same LOG_LEVEL-gated severity filtering) kept in sync here.
 *
 * LOG_LEVEL (debug|info|warn|error, default info) sets the minimum severity actually emitted -
 * DEBUG-level operational detail stays quiet by default so it doesn't crowd out the audit trail,
 * and is available on demand for troubleshooting without a code change.
 */
function minSeverity(): SeverityNumber {
  const level = process.env.LOG_LEVEL?.toLowerCase();
  return (level ? LEVELS[level] : undefined) ?? SeverityNumber.INFO;
}

function emit(
  severityNumber: SeverityNumber,
  severityText: string,
  eventName: string,
  attributes: AuditAttributes,
  error?: unknown,
): void {
  if (severityNumber < minSeverity()) {
    return;
  }
  const allAttributes: AuditAttributes = { ...attributes, [EVENT_NAME_ATTRIBUTE]: eventName };
  if (error !== undefined) {
    allAttributes['error.type'] = error instanceof Error ? error.constructor.name : typeof error;
    allAttributes['error.message'] = error instanceof Error ? error.message : String(error);
  }
  logs.getLogger(SCOPE).emit({ severityNumber, severityText, body: eventName, attributes: allAttributes });
}

/** Operational/debugging detail (e.g. a cache hit) - quiet unless LOG_LEVEL=debug. */
export function debug(eventName: string, attributes: AuditAttributes = {}): void {
  emit(SeverityNumber.DEBUG, 'DEBUG', eventName, attributes);
}

/** A business-level audit event. */
export function event(eventName: string, attributes: AuditAttributes = {}): void {
  emit(SeverityNumber.INFO, 'INFO', eventName, attributes);
}

/** Something recoverable went wrong or looks off (e.g. a fallback kicked in). */
export function warn(eventName: string, attributes: AuditAttributes = {}): void {
  emit(SeverityNumber.WARN, 'WARN', eventName, attributes);
}

/** A business-level audit event that failed. */
export function error(eventName: string, attributes: AuditAttributes, err: unknown): void {
  emit(SeverityNumber.ERROR, 'ERROR', eventName, attributes, err);
}
